hydroflow/src/builder/hydroflow_builder.rs

```diff
@@ -0,0 +1,124 @@
+use super::build::{PullBuild, PushBuild};
+use super::connect::{PullConnect, PushConnect};
+use super::surface::pivot::PivotSurface;
+
+use std::rc::Rc;
+use std::sync::mpsc::SyncSender;
+
+use crate::compiled::pivot::Pivot;
+use crate::scheduled::graph::Hydroflow;
+use crate::scheduled::graph_ext::GraphExt;
+use crate::scheduled::handoff::{CanReceive, Handoff};
+use crate::scheduled::input::Input;
+
+use super::surface::pull_handoff::HandoffPullSurface;
+use super::surface::push_handoff::HandoffPushSurfaceReversed;
+use super::surface::push_start::StartPushSurface;
+use super::surface::{PullSurface, PushSurfaceReversed};
+
+/// The user-facing entry point for the Surface API.
+#[derive(Default)]
+pub struct HydroflowBuilder {
+    hydroflow: Hydroflow,
+    // TODO(mingwei): use a dedicated trait instead of FnOnce?
+    handoff_connectors: Vec<Box<dyn FnOnce(&mut Hydroflow)>>,
+}
+impl HydroflowBuilder {
+    /// Creates a handoff, returning push and pull ends which can be chained
+    /// using the Surface API.
+    pub fn make_handoff<H, T>(
+        &mut self,
+    ) -> (HandoffPushSurfaceReversed<H, T>, HandoffPullSurface<H>)
```

Comment:
Should we expose `Surface` in the name for the top-level stuff? Maybe the interface should just be like `PushHandoff` `PullHandoff` or something and the inner layer can be `PushHandoffInner`/`PullHandoffInner`?

Reply:
Maybe in the future, for now I want to keep the naming clearer for us rather than the users, since this bit of code is kinda gnarly